### ISSUE TITLE
new rule proposal for arrow function

### ISSUE THREAD

**Issue by Jxck** [eslint/eslint#2628]
hi, I'm using my own plugin for arrow function.

https://github.com/Jxck/eslint-plugin-arrow-function

this works fine for me so i'm suggest to merging this for default rules.
any comments are welcome.

detail from readme is below.

---

### parens

The following patterns are considered warnings:

```js
a => {}
a => a
a => {\n}
a.then(foo => {});
a.then(foo => a);
a(foo => { if (true) {}; });
```

The following patterns are not warnings:

```js
() => {}
(a) => {}
(a) => a
(a) => {\n}
a.then((foo) => {});
a.then((foo) => { if (true) {}; });
```

this saves you from bizarre behavior like below

```js
var a = 1;
if (a => 2) {
 console.log('bigger');
} else {
 console.log('smaller')
};
```

this is better, because condition of if is arrow function, not comparison.
this should be like this, and you can notice it's not what you expect.

```js
var a = 1;
if ((a) => 2) {
 console.log('bigger');
} else {
 console.log('smaller')
};
```

same thing happens here.

```js
var a = 1, b = 2, c = 3, d = 4;
var f = a => b ? c: d;
// f = ?
```

`f` is arrow function which gets a as arguments and returns the result of `b ? c: d`.

this should be like this again.

```js
var a = 1, b = 2, c = 3, d = 4;
var f = (a) => b ? c: d;
```

you may notice what this is.

### no-condition

Disallow arrow function at places where a condition is expected.

The following patterns are considered warnings:

```js
if (a => 1) {}
while (a => 1) {}
for (var a = 1; a => 10; a++) {}
a => 1 ? 2 : 3
(a => 1) ? 2 : 3
```

Even if the arguments of the arrow function are wrapped with parens, this rule warns you about it.

### space

The following patterns are considered warnings:

```js
()=> {}
() =>{}
(a)=> {}
(a) =>{}
a =>a
a=> a
()=> {\n}
() =>{\n}
```

The following patterns are not warnings:

```js
() => {}
(a) => {}
a => a
() => {\n}
```

## Usage

```yaml
plugins:
  - arrow-function

rules:
  # Plugins
  arrow-function/parens : 2
  arrow-function/no-condition : 2
  arrow-function/space  : 2
```

**Comment by nzakas** [comment]
Given that you already have a plugin, why not just keep using the plugin?

**Comment by ilyavolodin** [comment]
I can see how `parens` and `space` can be useful to wider audience of users, but I think `no-condition` is too situational and specific to be considered for adding to the core.

**Comment by Jxck** [comment]
@nzakas because I think it's a basic rules for arrow function, which can be built in to core.

@ilyavolodin `no-condition` seems part of plugin is make sense. in that case. I'll send a PR only `parens` & `space`.

**Comment by hzoo** [comment]
:+1: to `parens` (kind of like `curly`) and `space` (similar to `space-before-function-paren`). I'm adding the `parens` rule to JSCS too

**Comment by nzakas** [comment]
Okay, can someone propose rule names?

**Comment by hzoo** [comment]
Maybe `arrow-parens`: with `always` and `never` and `arrow-spacing` with `before`, `after`, `both`, `never`?

**Comment by ilyavolodin** [comment]
I would change the last option for `arrow-spacing` to `none` instead of `never`, but other then that, those names look good to me

**Comment by Jxck** [comment]
+1 for @ilyavolodin

**Comment by mysticatea** [comment]
FYI: [generator-star-spacing](http://eslint.org/docs/rules/generator-star-spacing)'s option is `before`, `after`, `both`, or `neither`.

**Comment by hzoo** [comment]
Hmm I guess we can go with either of those - and maybe try to make all the `spacing-x` rules consistent in 1.0?

**Comment by nzakas** [comment]
We need to move fast if we want a change in 1.0. Which format do people prefer?

**Comment by ilyavolodin** [comment]
I think `{"before": true, "after": true}` is the best structure. It's the easiest to configure and the least verbose. I think we might want to create a more generic task to compare options among all rules and bring them to similar standard. Now that we have rule schemas it should be easy to get a list of all available options.

**Comment by mysticatea** [comment]
I also prefer `{"before": true, "after": true}`.

FYI, about other `*-spacing` rules:

- [key-spacing](http://eslint.org/docs/rules/key-spacing)'s option is an object such as `{"beforeColon": true, "afterColon": true}`. This rule has more an option for vertical alignments, so it's not `property-colon-spacing`.
- [object-curly-spacing](http://eslint.org/docs/rules/object-curly-spacing)'s option is `"never"` or `"alwas"`. Probably all `*-spacing` rules for parentheses are going to have the same option.

**Comment by Jxck** [comment]
I wanna clarify about `{ "before": true, "after": true }`.

eslintrc maybe looks like

```
"rules": {
  "arrow-spacing": [ 2, { "before": true, "after": true } ]
}
```

and default value is `{ "before": true, "after": true }`

so

`{ "before": true, "after": true }`
- 1 space around arrow like `() => {}` is valid.

`{ "before": false, "after": false }`
- no space around arrow like `()=>{}` is valid.

`{ "after": false }`
- same as `{ "before": true, "after": false }`

`{ "before": false }`
- same as `{ "before": false, "after": true }`

`"arrow-spaceing": [2, {}]` or only `"arrow-spaceing": 2`
- using default value

is that correct ?

**Comment by ilyavolodin** [comment]
@Jxck Yes, those examples are correct.

**Comment by Jxck** [comment]
@ilyavolodin ok, I implemented that in my code as PoC.
https://github.com/Jxck/eslint-plugin-arrow-function (v0.0.3)

**Comment by rightaway** [comment]
+1 for adding to core, plugin works great

**Comment by Jxck** [comment]
I sent a PR based on discussion here.
please review and any comments are welcome.
